File: js/packages/app/src/__tests__/features/object-references/external/wrappers/object-references-example/src/polywrap/wrapped/TestExternalClass/invoke.ts
import { parse } from '@serial-as/json'
import { bufferToU32, BaseTypeSerialization, IExternalWrapInstance } from "@nerfzael/reim-wrap-as";
import { WrapManifest } from '../../WrapManifest';
import { TestExternalClassWrapped } from "..";
import { TestExternalClass } from "../../external/classes";



export function invoke(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer {
  const funcId = bufferToU32(buffer);
  const dataBuffer = buffer.slice(4);

  switch (funcId) {
    case WrapManifest.Internal.Classes.TestExternalClassMethod.Create:
        return invokeCreateWrapped(dataBuffer, wrapInstance);
    case WrapManifest.Internal.Classes.TestExternalClassMethod.TestInstanceMethod:
        return invokeTestInstanceMethodWrapped(dataBuffer, wrapInstance);
    case WrapManifest.Internal.Classes.TestExternalClassMethod.TestStaticMethod:
        return invokeTestStaticMethodWrapped(dataBuffer, wrapInstance);
    default:
      throw new Error(`Unknown method: ${funcId.toString()} on class `);
  }
}

const invokeCreateWrapped = (buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer => {
  const args = CreateArgsWrapped.deserialize(buffer, wrapInstance);

  const result = TestExternalClass.create(
    args.arg,
  );
  
  return TestExternalClassWrapped.serialize(result);
  };

class CreateArgs {
  constructor(
    public arg: string,
  ) {
  }
}

@serializable
class CreateArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): CreateArgs {
    const args = parse<CreateArgsWrapped>(String.UTF8.decode(buffer));
  
    return new CreateArgs(
            
      args.arg,
      
    );
  }  
}
const invokeTestInstanceMethodWrapped = (buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer => {
  
  const referencePtr = bufferToU32(buffer);
  const dataBuffer = buffer.slice(4);

  const args = TestInstanceMethodArgsWrapped.deserialize(dataBuffer, wrapInstance);

  const object = TestExternalClassWrapped.dereference(referencePtr);

  const result = object.testInstanceMethod(
    args.arg,
  );

  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestInstanceMethodArgs {
  constructor(
    public arg: string,
  ) {
  }
}

@serializable
class TestInstanceMethodArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): TestInstanceMethodArgs {
    const args = parse<TestInstanceMethodArgsWrapped>(String.UTF8.decode(buffer));
  
    return new TestInstanceMethodArgs(
            
      args.arg,
      
    );
  }  
}
const invokeTestStaticMethodWrapped = (buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer => {
  const args = TestStaticMethodArgsWrapped.deserialize(buffer, wrapInstance);

  const result = TestExternalClass.testStaticMethod(
    args.arg,
  );
  
  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestStaticMethodArgs {
  constructor(
    public arg: string,
  ) {
  }
}

@serializable
class TestStaticMethodArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): TestStaticMethodArgs {
    const args = parse<TestStaticMethodArgsWrapped>(String.UTF8.decode(buffer));
  
    return new TestStaticMethodArgs(
            
      args.arg,
      
    );
  }  
}

File: rust/app/src/wrappers/external_references/src/polywrap/external/wrapped/test_external_class/invoke.rs
use reim_wrap::{ ExternalModule };
use module::{ WrapManifest };
use internal::classes::{ TestExternalClassWrapped };
use external::classes::{ TestExternalClass };



pub fn invoke(buffer: &[u8], external_module: dyn ExternalModule) -> Vec<u8> {
  let func_id = u32::from_be_bytes(buffer.try_into().expect("Method ID must be 4 bytes"));
  let data_buffer = &buffer[4..];

  match func_id {
    wrap_manifest::internal::classes::TestExternalClassMethod::Create =>
        invokeCreateWrapped(data_buffer, external_module),
    wrap_manifest::internal::classes::TestExternalClassMethod::TestInstanceMethod =>
        invokeTestInstanceMethodWrapped(data_buffer, external_module),
    wrap_manifest::internal::classes::TestExternalClassMethod::TestStaticMethod =>
        invokeTestStaticMethodWrapped(data_buffer, external_module),
    _ => panic!(format!("Unknown method: {} on class {}", func_id.to_string(), model.class_name))
  }
}

fn invokeCreateWrapped(buffer: &[u8], external_module: dyn ExternalModule) -> Vec<u8> {
    let args = CreateArgsWrapped.deserialize(buffer, external_module);

    let result = TestExternalClass.create(
        args.arg,
    );
    
    TestExternalClassWrapped.serialize(result)
    };

struct CreateArgs {
    pub arg: String,
}

impl CreateArgs {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }
}

struct CreateArgsWrapped {
    pub arg: String,
}

impl CreateArgsWrapped {
    pub fn new(
        arg: String,
    ) {
        Self {
        arg,
        }
    }

    pub fn deserialize(buffer: Uint8Array, external_module: dyn ExternalModule) -> CreateArgs {
        let args = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
    
        CreateArgs.new(
                
        args.arg,
        
        )
    }  
}
fn invokeTestInstanceMethodWrapped(buffer: &[u8], external_module: dyn ExternalModule) -> Vec<u8> {
    
    let reference_ptr = u32::from_be_bytes(buffer.try_into().expect("Reference ptr must be 4 bytes"));
    let data_buffer = &buffer[4..];

    let args = TestInstanceMethodArgsWrapped.deserialize(data_buffer, external_module);

    let object = TestExternalClassWrapped.dereference(reference_ptr);

    let result = object.testInstanceMethod(
        args.arg,
    );

    
    BaseTypeSerialization.serialize<String>(result)
};

struct TestInstanceMethodArgs {
    pub arg: String,
}

impl TestInstanceMethodArgs {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }
}

struct TestInstanceMethodArgsWrapped {
    pub arg: String,
}

impl TestInstanceMethodArgsWrapped {
    pub fn new(
        arg: String,
    ) {
        Self {
        arg,
        }
    }

    pub fn deserialize(buffer: Uint8Array, external_module: dyn ExternalModule) -> TestInstanceMethodArgs {
        let args = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
    
        TestInstanceMethodArgs.new(
                
        args.arg,
        
        )
    }  
}
fn invokeTestStaticMethodWrapped(buffer: &[u8], external_module: dyn ExternalModule) -> Vec<u8> {
    let args = TestStaticMethodArgsWrapped.deserialize(buffer, external_module);

    let result = TestExternalClass.testStaticMethod(
        args.arg,
    );
    
    
    BaseTypeSerialization.serialize<String>(result)
};

struct TestStaticMethodArgs {
    pub arg: String,
}

impl TestStaticMethodArgs {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }
}

struct TestStaticMethodArgsWrapped {
    pub arg: String,
}

impl TestStaticMethodArgsWrapped {
    pub fn new(
        arg: String,
    ) {
        Self {
        arg,
        }
    }

    pub fn deserialize(buffer: Uint8Array, external_module: dyn ExternalModule) -> TestStaticMethodArgs {
        let args = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
    
        TestStaticMethodArgs.new(
                
        args.arg,
        
        )
    }  
}
